[TASK-6946]: fix fee calculation in cashout and request flows

kushagrasarathe opened the pull request:
- added new info row component to render row's in the view
- updated calculations for slippage, gas and total

fixes TASK-8331 TASK-6946

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enhanced transaction fee and slippage calculations for cross-chain payments.
	- Introduced new components (`FeeDescription` and `InfoRow`) for displaying detailed transaction fee information.
	- Improved error handling and state management for payment requests.
	- Enhanced UI presentation for transaction costs using the new `InfoRow` component.
	- Added a new submodule for animations to improve visual elements.
	- Included `estimatedGasCost` in various components for better financial detail management.

- **Bug Fixes**
	- Refined transaction preparation and estimation logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kushagrasarathe committed:
feat: add maxTotal for request pay + slippage and network cost range

kushagrasarathe committed:
feat: update max total to range + info row ui component

Hugo0 commented on src/components/Request/Pay/Views/Initial.view.tsx:
```diff
@@ -97,16 +96,27 @@ export const InitialView = ({
+        const gasEstimate = estimatedGasCost ?? 0
+        return {
+            min: formatTokenAmount(gasEstimate, 3) ?? '0',
+            max: formatTokenAmount(gasEstimate * 1.1, 3) ?? '0',
```
Q & issues, blocking:
1. Where do these super arbitrary values come from? 
2. If we're going to  have gas consts multipliers like this, then these should probably not be defined inline in this page
3. we shouldn't have custom code for fees in request pay, and not have this in claim create, claim, and cashout, and request create. These all should share a common fees section!
4. where does estimatedGasCost come from? SDK? In that case, it already has a buffer included! Also, the _actual_ gas cost will be below the estimatedGasCost, sometimes as much as 50%

Hugo0 commented on src/components/Request/Pay/Views/Initial.view.tsx:
```diff
@@ -157,6 +185,32 @@ export const InitialView = ({
+                    min: formatAmount(minTotal),
+                    max: formatAmount(maxTotal),
+                })
+
```
pls always resolve coderabbit comments - mostly they're unnecessary nitpicks, but sometimes  they're useful

Hugo0 reviewed (changes requested):
I'd like to expand the scope of this PR to affect all flows and be consistent accross them

kushagrasarathe committed:
fix: fee calculations + add fee description component

kushagrasarathe committed:
fix: reuse info row component + add loading

kushagrasarathe committed:
fix: extract numbers as consts

kushagrasarathe committed:
fix: add git modules

kushagrasarathe committed:
fix: add improved loading conditions

kushagrasarathe committed:
fix: fee calculation in cashout flow

kushagrasarathe committed:
Merge branch 'develop' into fix/request-ux

kushagrasarathe committed:
fix: refactor :)

kushagrasarathe committed:
refactor: use claculate fee fn + improve ux copy

Hugo0 commented on src/components/Cashout/Components/Initial.view.tsx:
```diff
@@ -1,26 +1,26 @@
+import { useContext, useEffect, useMemo, useState } from 'react'
+import { twMerge } from 'tailwind-merge'
+import { useAccount } from 'wagmi'
+import * as _consts from '../Cashout.consts'
```
nit: better explicit imports

Hugo0 commented on src/components/Cashout/Components/Initial.view.tsx:
```diff
@@ -125,6 +126,25 @@ export const InitialCashoutView = ({
+                'unsignedTxs' in preparedCreateLinkWrapperResponse.response &&
+                preparedCreateLinkWrapperResponse.response.unsignedTxs[0]
+            ) {
+                try {
+                    const { transactionCostUSD } = await estimateGasFee({
+                        chainId: selectedChainID,
+                        preparedTx: preparedCreateLinkWrapperResponse.response.unsignedTxs[0],
+                    })
```
Q: What if there's two transactions? Don't we estimate the gas cost of the second one? Sometimes unsignedTxs can be 2 transactions, if an approval is needed. So the 1st transaction would be cheap, just an approval, whilst the 2nd transaction is expensive (swap/bridge)

kushagrasarathe replied:
@Hugo0 this condition on the initial cashout view is purely being used to check if the gas fee for the tx is less than the amount user wants to withdraw and set a proper error state if that's not the case. 
so imo this is fine

Hugo0 replied:
Wait so we're doing a whole RPC call and estimate gas fee, JUST to check if the `gas fee < total amount` as safety? That's dumb as hell then, we alrdy have a $10 minimum, and this introduces latency for everyone

If this is the case  we should delete this imo

Hugo0 commented on src/components/Cashout/Cashout.consts.ts:
```diff
@@ -1,9 +1,9 @@
+import { CreateScreens } from '../Create/Create.consts'
 import { OfframpConfirmView, OfframpSuccessView } from '../Offramp'
 import { OfframpType } from '../Offramp/Offramp.consts'
+import * as views from './Components'
```
nit: explicit is better

Hugo0 commented on src/components/Claim/Link/Initial.view.tsx:
```diff
@@ -1,43 +1,40 @@
+import * as consts from '@/constants'
 import { TOOLTIPS } from '@/constants/tooltips'
-import AddressLink from '@/components/Global/AddressLink'
-import TokenSelector from '@/components/Global/TokenSelector/TokenSelector'
-import { checkTokenSupportsXChain, SQUID_ETH_ADDRESS } from '@/utils/token.utils'
-import { getSquidTokenAddress } from '@/utils/token.utils'
+import * as context from '@/context'
+import { useAuth } from '@/context/authContext'
+import {
+    areTokenAddressesEqual,
+    checkifImageType,
+    ErrorHandler,
+    formatTokenAmount,
+    getBridgeChainName,
+    getBridgeTokenName,
+    saveClaimedLinkToLocalStorage,
+} from '@/utils'
+import { checkTokenSupportsXChain, getSquidTokenAddress, SQUID_ETH_ADDRESS } from '@/utils/token.utils'
+import { Popover } from '@headlessui/react'
+import { useAppKit } from '@reown/appkit/react'
+import { getSquidRouteRaw } from '@squirrel-labs/peanut-sdk'
+import { useContext, useEffect, useMemo, useState } from 'react'
+import { useAccount } from 'wagmi'
+import * as _consts from '../Claim.consts'
+import useClaimLink from '../useClaimLink'
```
Q: why so much change lol

kushagrasarathe replied:
my vs code workspace settings 😭
helps with formatting and removing unused imports

Hugo0 commented on src/components/Create/Link/Confirm.view.tsx:
```diff
@@ -293,26 +293,22 @@ export const CreateLinkConfirmView = ({
-                        </label>
-                    </div>
+                                  ? '$ 0.01'
+                                  : `$ ${formatTokenAmount(transactionCostUSD, 3) ?? 0}`
```
nit: this seems slightly messy

Hugo0 commented on src/components/Create/Link/Confirm.view.tsx:
```diff
@@ -293,26 +293,22 @@ export const CreateLinkConfirmView = ({
-                            />
-                        </label>
-                    </div>
+                                  ? '$ 0.01'
```
Q: Whats the reasoning to showing 1 cent min? I'd show `< $0.01` in that case

Also, blocking, we should be consistent with space and dollar sign

Hugo0 commented on src/components/Global/FeeDescription/index.tsx:
```diff
@@ -0,0 +1,152 @@
+                )}
+            </div>
+
+            {/* Expandable Section */}
```
Nit: design wise, I think the expandable section should be less "primary" and seem more secondary to the estimated fee section

Hugo0 replied:
so, i'd create a secondary info row or similar

Hugo0 commented on src/components/Global/FeeDescription/index.tsx:
T: I'm wondering if it's not better to have the fee logic in one central place (ie this component), not just the display.

Advantages: No duplicate fee calc bugs
Disadvantages: Not sure the abstraction can be neatly built

kushagrasarathe replied:
have a task, will prioritize it and update acoordingly

Hugo0 commented on src/components/Global/InfoRow/index.tsx:
nice

Hugo0 commented on src/components/Offramp/Confirm.view.tsx:
```diff
@@ -646,6 +654,53 @@ export const OfframpConfirmView = ({
+        if (hasPromoCode) return baseGasCost
+
+        const accountFee = accountType === 'iban' ? 1 : 0.5
+        const crossChainFee = hasCrossChain ? baseGasCost * 0.1 : 0
```
issue, blocking: I feel like this will be _really_ hard to reason about in 6 months. We need to link to discussion here or explain these params. Why are we multiplying by 0.1? What's the logic behind it? etc

Hugo0 replied:
Also, why are we multiplying the base gas cost x 0.1? It should be the slippage, not the gas cost

kushagrasarathe replied:
@Hugo0 its slippage only, we are using 10% of gas cost as slippage right, this is what we discussed

Hugo0 replied:
but its called `baseGasCost` - that sounds like gas cost, not slippage cost! Either the variable is misnamed, or we're mistaking the variable and multiplying gascost with 10%, when it should be 70%

Hugo0 commented on src/components/Request/Pay/Views/Initial.view.tsx:
```diff
@@ -98,22 +96,68 @@ export const InitialView = ({
 
+    const feeCalculations = useMemo(() => {
+        // percentage of maximum network fee to use as expected fee (70%)
+        const EXPECTED_NETWORK_FEE_MULTIPLIER = 0.7
+
+        // percentage of maximum slippage to use as expected slippage (10%)
+        const EXPECTED_SLIPPAGE_MULTIPLIER = 0.1
```
Thought, blocking: I think this is nice. But its sad that we only use this in request, and not in cashout or send.

I think we should centralize the fee logic in one place, else this is going to become a big source of bugs and user complaints

kushagrasarathe replied:
Agree, I'll take care of this thing in this week, and move the whole fee logic into a single place and reuse it, its a relatively bigger task, so I'll prioritize stuff

kushagrasarathe committed:
Merge branch 'develop' into fix/request-ux

kushagrasarathe committed:
fix: fee calculation for cashout + pr req changes

kushagrasarathe committed:
refactor: update to use expected fee instead of max

kushagrasarathe committed:
fix: implement cashout feedback
- update icons
- implement visual difference for dropdown items
- implement cashout feedback

Hugo0 commented on src/components/Offramp/Confirm.view.tsx:
```diff
@@ -740,7 +740,7 @@ export const OfframpConfirmView = ({
             ) : (
                 <div className="flex w-full flex-col items-center justify-center gap-2">
                     <label className="self-start text-h8 font-light">Please confirm all details:</label>
-                    <div className="flex w-full flex-col items-center justify-center gap-2">
+                    <div className="flex w-full flex-col items-center justify-center gap-3">
```
nit: so much individual styling

Hugo0 replied:
gap-2 gap-3 etc

kushagrasarathe replied:
updated the gaps for ui to look better

Hugo0 commented on src/components/Create/Link/Confirm.view.tsx:
```diff
@@ -296,18 +296,20 @@ export const CreateLinkConfirmView = ({
                         iconName="gas"
                         label="Estimated Fee"
-                        value={`~ ${
+                        value={`${
```
nit: why remove the ~? i think its good

Hugo0 commented on src/components/Create/Link/Confirm.view.tsx:
```diff
@@ -296,18 +296,20 @@ export const CreateLinkConfirmView = ({
-                                  ? '$ 0.01'
-                                  : `$ ${formatTokenAmount(transactionCostUSD, 3) ?? 0}`
+                                  ? '~ $ 0.01'
+                                  : `~ $ ${formatTokenAmount(transactionCostUSD, 3) ?? 0}`
```
oh, we're adding it here

kushagrasarathe committed:
fix: resolve pr review comments

kushagrasarathe committed:
fix: use crosschain slippage percentage to clculate slippage on cashout flow

jjramirezn commented on src/components/Request/Pay/Views/Initial.view.tsx:
```diff
@@ -99,22 +98,71 @@ export const InitialView = ({
+        }
+
+        // slippage calculation (10% of max for expected)
+        const slippage = calculatedSlippage
```
issue: the slippage used will always be max

In the case of slippage in requests, we always use the max amount, because we have to ensure that the requester never receives less than requested.

kushagrasarathe replied:
@jjramirezn  updated the ui,

total max = amount + gas + max slippage
network cost = max network cost
estimated fee = expected network cost + max slippage

<img width="414" alt="Screenshot 2025-01-29 at 4 39 06 PM" src="https://github.com/user-attachments/assets/26dc7a5d-252d-453c-828b-6e0ff15e5a0e" />

kushagrasarathe replied:
@jjramirezn  reminder to check this

kushagrasarathe committed:
fix: use max slippage in req flow for fee calculation

kushagrasarathe committed:
fix: loading state + render max slippage on pay view

kushagrasarathe committed:
fix(style): dollar spacing 💰
